### DISCUSSION TITLE
第三回 MTG (9/12)

### DISCUSSION THREAD

**Discussion by MachineHunter** [MachineHunter/SafeExecute#14] [General]
## 話し合いたい内容
* このツールの全体の仕組みのおさらい
* 作業の流れがどんな感じになるかのイメージの共有
* どういう風にユーザーが禁止する挙動を設定できるようにするか
* GUI化するかどうか、する場合のレイアウト
* そもそも皆さんどの部分の開発に興味があるか（UI、検知実装、リサーチ みたいな）
* 現状悪性っぽい挙動が見つかったら `ExitProcess` という方針だが、プロセス終了以外どういう機能があったら嬉しいか
* このツールのメリット/デメリットをそれぞれ挙げられるだけあげる
* 検証/テスト方法（クリーンウェアとマルウェアを持ってきて色々試すとか）

## 話し合った内容
* できればGUIがいい
* 一括チェックができる感じのツリー型のチェックボックス
* 役割分担（自分:適当、UI:2、検知:3）
  * 小林: どれもやりたい
  * 杉山: 検知実装/リサーチ
  * 林: UI
  * Shupei: どれもやりたい
  * 丹治: 
* プロセス終了以外の機能（ツールの機能案）
  * どの挙動を検知してExitProcessしたかを表示する
  * 一時停止して、例えば「こういうURL(引数)にアクセスしようとしてますが、実行続行しますか？」みたいなのを表示し、続行 or 終了 できるようにする
  * 制限した挙動を発見した場合、終了するか一時停止するかをモードで切り替えられるようにする
  * スキップ機能（そのWindowsAPIを実行せずに、後続の処理を続ける）（できるものは）
  * 今まで検知されたWindowsAPIを集計する（とりあえずローカルのcsvファイルとかに保存する）（保存するかどうかもモードで選べるようにするといいかも）
  * どうせフックの作業似たような感じなので、動的にフックを追加できるようにして、チェックボックスに無いWindowsAPIもフックできるようにする（**僕の宿題**）
* このツールのメリット
  * ユーザーが実行を制御できる
  * 正常なバイナリだった場合に、正常な実行に全く支障が無いです
  * 軽い
    * アンチウイルスだとアップデートで処理が重くなったりするけど、そういうのに比べて軽量（容量）
    * バックグラウンドで常時起動してるわけではなく、オンデマンドで実行できる
  * アンチウイルスだと動かしたいプログラムを勝手に消されたりするが、そんなことはしません！
  * 容量
* このツールのデメリット
  * Windows API をあらかじめ知っておく必要があるんじゃないか？
  * フックできるWinAPIがチェックボックスの項目だけになってしまうので、指定したい項目が無いとツールが全く使い物にならなくなってしまう
    * => 検知実装/リサーチ の方により力を入れる必要がありそう（できるだけ多くのWinAPIを網羅できるようにした方が良さげ）
 * 検証/テスト方法
   * 「正常な実行に全く支障が無い」これに関しては、でかいゲームの実行ファイルとかが正常に動く事とかを試せばいい
   * クロックサイクル数計測するプログラムとかで、フック有り/無しで時間の差を見てみる
   * 実際のマルウェアとかでもいくつか試してみた方がいいかも

## 次のMTGまでにやる事
* 検知実装/リサーチ班（+現状どれもやりたいの人）
  * issueにあるタスクをフックの練習的な感じでやってみる
  * 終わったら、どういうWindowsAPIがどういう引数で呼ばれてるとマルウェアっぽいのかとかを、[このissue](https://github.com/MachineHunter/SafeExecute/issues/15)で追加したテスト用バイナリとかが使ってるWinAPIとかを参考にしながら考えてみて、それを検知する仕組みをissueとして発行して実装したりみたいな感じで進めていただければ
* GUI班
  * SafeExecuteに新たにWinFormプロジェクトを作成し、Panelでの区画分けからやってみる
  * それが終わったら、一旦相談しましょう!ｗ

## その他
* 何か追加とか相談内容とかあればどんどんDiscussionsのコメントに追加していってください！（github荒らしてくださいｗ）
*

**Comment by KOHsEMP** [comment]
どの挙動を検知してExitProcessしたかを表示する(MessageBox ?) のはどうでしょうか？
